core/lib/zksync_core/src/state_keeper/state_keeper_storage.rs

```diff
@@ -0,0 +1,249 @@
+                        snapshot_recovery.l1_batch_number,
+                    )
+                }
+            };
+
+        tracing::debug!(%l1_batch_number, %miniblock_number, "Using Postgres-based storage");
+        Ok(
+            PostgresStorage::new_async(rt_handle, connection, miniblock_number, true)
+                .await?
+                .into(),
+        )
+    }
+
+    /// Catches up RocksDB synchronously (i.e. assumes the gap is small) and
+    /// returns a [`ReadStorage`] implementation backed by caught-up RocksDB.
+    async fn access_storage_rocksdb<'a>(
+        conn: &mut StorageProcessor<'_>,
+        rocksdb: RocksDB<StateKeeperColumnFamily>,
+        stop_receiver: &watch::Receiver<bool>,
+    ) -> anyhow::Result<Option<PgOrRocksdbStorage<'a>>> {
+        tracing::debug!("Catching up RocksDB synchronously");
+        let rocksdb_builder: RocksdbStorageBuilder = rocksdb.into();
+        let rocksdb = rocksdb_builder
+            .synchronize(conn, stop_receiver)
+            .await
+            .context("Failed to catch up state keeper RocksDB storage to Postgres")?;
+        let Some(rocksdb) = rocksdb else {
+            tracing::info!("Synchronizing RocksDB interrupted");
+            return Ok(None);
+        };
+        let rocksdb_l1_batch_number = rocksdb.l1_batch_number().await.unwrap_or_default();
+        tracing::debug!(%rocksdb_l1_batch_number, "Using RocksDB-based storage");
+        Ok(Some(rocksdb.into()))
+    }
+
+    async fn access_storage_inner<'a>(
+        &'a self,
+        rt_handle: Handle,
+        stop_receiver: &watch::Receiver<bool>,
+    ) -> anyhow::Result<Option<PgOrRocksdbStorage<'a>>> {
+        match self.rocksdb_cell.get() {
+            Some(rocksdb) => {
+                let mut conn = self
+                    .pool
+                    .access_storage_tagged("state_keeper")
+                    .await
+                    .context("Failed getting a Postgres connection")?;
+                Self::access_storage_rocksdb(&mut conn, rocksdb.clone(), stop_receiver)
+                    .await
+                    .context("Failed accessing RocksDB storage")
+            }
+            None => Ok(Some(
+                Self::access_storage_pg(rt_handle, &self.pool)
+                    .await
+                    .context("Failed accessing Postgres storage")?,
+            )),
+        }
+    }
+
+    pub fn new(
+        pool: ConnectionPool,
+        state_keeper_db_path: String,
+        enum_index_migration_chunk_size: usize,
+        stop_receiver: watch::Receiver<bool>,
+    ) -> Self {
+        let rocksdb_cell = Arc::new(OnceCell::new());
+        let task = AsyncCatchupTask {
+            pool: pool.clone(),
+            state_keeper_db_path,
+            enum_index_migration_chunk_size,
+            rocksdb_cell: rocksdb_cell.clone(),
+        };
+        tokio::task::spawn(async move { task.run(stop_receiver).await.unwrap() });
```

Comment:
I'd return `AsyncCatchupTask` from this method so that the caller can manage it. E.g., it'd make sense to add it to the list of managed tasks on the main node / EN (rn you'd need to call [`pending().await`](https://doc.rust-lang.org/std/future/fn.pending.html) so that when the task is logically complete, it doesn't bring the node down with it).

Reply:
I used `stop_receiver.0.changed().await?` as that seemed like a slightly better alternative to `pending().await`